### ISSUE TITLE
Antipattern: nix-channels

### ISSUE THREAD

**Issue by domenkozar** [domenkozar/nix.dev#16]

**Comment by samueldr** [comment]
This was linked [in an answer about "channels being phased out"](https://discourse.nixos.org/t/channels-being-phased-out/7836).

There was recently a user asking questions on the NixOS IRC channel, confused about what "channels being phased-out" meant.

As this article is part of the website, this it not exactly great. This is giving out a message that officially something is decided here, while I don't even know where to start looking for a citation about that.

Furthermore, the link "[Channels](https://nixos.wiki/wiki/Nix_channels)" points to the NixOS wiki article that is more about the *Hydra process of marking revisions as good and tested* than the `nix-channel` command (which I assume is what you state is being phased out).

What was the meaning of that note? Is Hydra being phased out? (I guess not :)) Is it about `nix-channel` being *de facto* being replaced with Flakes? Is there more?

**Comment by domenkozar** [comment]
Until I have time to write about channels, beginners have to blindly believe they are better off not using channels.

The constraint here is my time, but this is OSS project so everyone is welcome to contribute :)

**Comment by samueldr** [comment]
Yes, I understand it's OSS and everyone is welcome to contribute.

The problem I have is I don't know where you are going off with that thought.

This is currently used on the official website. So this leaves kind of a weird unfinished thought about a tool that is part of the tool, and distro! There is no announcement of the sort, but it's stated that it's being phased out, but nothing about what replaces it.

cc @garbas

**Comment by domenkozar** [comment]
It's possible to avoid channels without flakes, although flakes will make that a lot simpler.

https://nix.dev/reference/pinning-nixpkgs.html offers solutions, although it's not complete nor structured well.

**Comment by domenkozar** [comment]
So I'm going to collect here all the reasons channels are an anti-pattern:

- it's an abstraction that provides CRUD operations over an archive (usually zip/tarball) which is mapped to a commit, but provide no benefits over that abstraction

- it's hard to inspect what git revision is in an archive (nixpkgs has a function to retrieve that, but in general case of channels there's no way to check what version of the channel you're using)

- it's global state that impacts the whole Nix evaluation results, defeating the whole purpose Nix was created for (to have deterministic properties for packaging)

- the implementation has some really weird bugs: [1](https://github.com/NixOS/nix/issues/4110) [2](https://github.com/NixOS/nix/issues/1211) [3](https://github.com/NixOS/nix/issues/2675)

- [did I say it has some really weird UX bugs?](https://github.com/NixOS/nix/issues/1548)

- it's possible to [inherit channels between root and user](https://github.com/NixOS/nix/issues/2980), which is confusing even to experienced Nix users

- my pet peeve: it doesn't validate input for commands and provides really poor UX for newcomers

---

So I see two simple alternatives that exists today by managing Nix sources via source control:

a) Either use `builtins.fetchTarball("channel:nixos-20.09")` and always get the latest release (I wish it prefix was named release, not channel).

b) Or you manage nixpkgs source yourself inside git, by either using niv of flakes

**Comment by roberth** [comment]
* Widespread conflation of nix-channel, NIX_PATH and the hydra.nixos.org channels, all of which _can_ but probably shouldn't be called channels.

 * Relies on NIX_PATH which fundamentally does not support composition of expressions that need a different value for it, in contrast to function arguments or even Nix flakes' dependency injection.


> it's global state that impacts the whole Nix evaluation results

Even worse, machine-or-user-or-both-global. If only it was global; you ssh somewhere and your channel may as well not exist. Am I thinking too big? :)

> builtins.fetchTarball("channel:nixos-20.09") and always get the latest release (I wish it prefix was named release, not channel).

It appears like only `hydra.nixos.org` will be able to provide "channels" in the newspeak?

**Comment by nat-418** [comment]
> So I'm going to collect here all the reasons channels are an anti-pattern:
> 
>     * it's an abstraction that provides CRUD operations over an archive (usually zip/tarball) which is mapped to a commit, but the abstraction provides no benefits over just using pinning a git revision
> 
>     * it's hard to inspect what git revision is in an archive (nixpkgs has a function to retrieve that, but in general case of channels there's no way to check what version of the channel you're using)
> 
>     * it's global state that impacts the whole Nix evaluation results, defeating the whole purpose Nix was created for (to have deterministic properties for packaging)
> 
>     * the implementation has some really weird bugs: [1](https://github.com/NixOS/nix/issues/4110) [2](https://github.com/NixOS/nix/issues/1211) [3](https://github.com/NixOS/nix/issues/2675)
> 
>     * [did I say it has some really weird UX bugs?](https://github.com/NixOS/nix/issues/1548)
> 
>     * it's possible to [inherit channels between root and user](https://github.com/NixOS/nix/issues/2980), which is little known even to experienced Nix users, like which one overrides which?
> 
>     * my pet peeve: it doesn't validate input for commands and provides really poor UX for newcomers
> 
> 
> So I see two simple alternatives that exists today by managing Nix sources via source control:
> 
> a) Either use `builtins.fetchTarball("channel:nixos-20.09")` and always get the latest release (I wish it prefix was named release, not channel).
> 
> b) Or you let git version nixpkgs source, by either: pinning the revision or managing its lifecycle using niv or flakes

My understanding is that channels have nothing to do with version control systems per se, and so the parts of your critique related to the "needless abstraction" that is "mapped to a commit" are not pertinent. Nix is older than git,
and does not (or at least should not) have a hard git dependency.

**Comment by ntninja** [comment]
I definitely prefer channels over `fetchTarball "https://github.com/NixOS/nixpkgs/archive/nixos-25.11.tar.gz"` (or similar).

With channels, I can control when the system will potentially start evaluating/downloading new software by virtue of having to explicitly enter the `nix-channel --update`  command.

With the `fetchTarball`/flake way, the Nix commands may at any time decide that *now* is a great to interrupt my work with blocking on a slow 40MiB download of the latest Nix expressions followed by potentially downloading further updates from Nix cache or starting to rebuild some software for the next half hour if there is a cache miss.

Without a way to decide *when* nix tools will update referenced sources, `channels` still have the massively better UX in that “don’t interrupt my workflow” way.

**Comment by KiaraGrouwstra** [comment]
> It's possible to avoid channels without flakes, although flakes will make that a lot simpler.
> 
> https://nix.dev/reference/pinning-nixpkgs.html offers solutions, although it's not complete nor structured (yet) well.

it could be helpful to e.g. mention [jade's article](https://jade.fyi/blog/pinning-nixos-with-npins/) on this (depending on the outcome of #1194)

edit: addressed in #1198

**Comment by ntninja** [comment]
I see, [npins](https://github.com/andir/npins) would definitely fix the “it randomly downloads things when you least need it” problem. ✔️
